### DISCUSSION TITLE
Solution if endpoints are not being created / Getting 404 Responses

### DISCUSSION THREAD

**Discussion by AndresCarignano** [TCattd/HTMX-API-WP#9] [General]
So I've been struggling for about an hour with this but found a solution.

add_rewrite_endpoint() was not doing it's thing so every request to the /wp-htmx/v1/* was returning 404.

This had never happened in other websites but this installation is particulary bloated with plugins. Hard to say which one is messing with flush_rewrite_rules() but it prevented the urls to be remapped. My suspicions are Elementor Pro or Polylang Pro.

**Dirty solution:**
Go to Permalinks page under Settings > Permalinks and click **Save**.

[Found it here, for more info](https://wordpress.stackexchange.com/questions/167381/add-rewrite-endpoint-not-working)

**Comment by EstebanForge** [comment]
That's weird.

Elementor, as far as i know, should mess up with WP urls.
Polylang in general, yes. That's what those plugins are meant to do. Just like WPML o Bogo. But it's weird, because this endpoints are new. They shouldn't be touched by other plugins at all.

Also, this plugin (HTMX-API-WP) does a flush_rewrite_rules() on activation. See HXWP_Activate_Deactivate() class.

I don't have access to WPML nor Polylang Pro. But i could test this with the free version of Polylang and also Bogo if that helps.

Thanks for the heads up, @AndresCarignano

**Reply by AndresCarignano** [reply]
Yes HXWP_Activate_Deactivate() does all it can.

But this is not the only problem I've had with permalinks so I think something is breaking the flush_rewrite_rules() function itself here... The only way to flush correctly is manualy via the admin panel. Weird bug.

If you can't recreate it I would just add a disclaimer in the readme.

**Comment by jasalt** [comment]
There's something weird with the custom rewrite endpoints and WP permalinks and I don't think it has anything to do with other plugins. I noted similar issue with another plugin doing pretty much the same thing https://github.com/svandragt/htmxpress/issues/10.

With this plugin, after installation and copying the example `htmx-templates/htmx-demo.htmx.php` to my theme directory, hx queries to `/wp-htmx/v1/htmx-demo` were giving simply 404. I diagnosed with XDebug and the plugin was returning from execution at:

https://github.com/TCattd/HTMX-API-WP/blob/cdfb0e03db9073618a7ec071e3a0b8514ab0439f/classes/class-hxwp-render.php#L38

Reading the plugin code, I continued fiddling with my theme code and was able to get the correct HTML content as response to the browser, but it still oddly came as 404 response and didn't do the swap, by doing hx query `/wp-htmx/v1/htmx-demo?wp-htmx=htmx-demo` and just `/wp-htmx/v1/?wp-htmx=htmx-demo` . So I guessing that only the url parameter gets registered during the plugin activation automatically or something. Maybe that could help some figuring it out..

Currently to get the plugin working, permalink settings have to be changed/re-saved by hand after plugin activation, after that the it seems to work as advertised in the repository readme.

**Reply by EstebanForge** [reply]
@jasalt thanks for the input here.

It's weird. The plugin does include an activation routine that fires and flushes WP permalink structure:
https://github.com/TCattd/HTMX-API-WP/blob/cdfb0e03db9073618a7ec071e3a0b8514ab0439f/classes/admin/class-hxwp-activation.php

I just tested that locally, and it works...

https://i.ibb.co/fD72jZP/Pasted-image.png

That method is called when the plugin is activated. Rewrite rules should be flushed then.

**Reply by jasalt** [reply]
Yes, seems that it does not work universally, htmxpress behaves similar way with this also https://github.com/svandragt/htmxpress/blob/main/htmxpress.php#L23.

For me, I had to toggle permalink setting manually on both local VVV Virtualbox installation, and with WP installation on a server.

**Reply by EstebanForge** [reply]
I don't know why this happens, and why i haven't being able to reproduce it either. Not live or local.

I can think of a way to "patch" this, hooking into the plugin options, and flush permalink when the user saves the settings.

But that only patch this issue and minimizes the occurrence of that bug (in theory).
